### DISCUSSION TITLE
Atomic virial calculation in lammps

### DISCUSSION THREAD

**Discussion by pobo95** [ACEsuit/mace#272] [Ideas]
First of all, thank you for making a good code.

I recently study MACE and just test MD simulation in lammps.

Unfortunately, I found that it doesn't support vflag_atom which for atomic virial stress calculation.

I'm not farmiliar with cpp file but 

 if(vflag_atom) {
    torch::Tensor atomic_virial_tensor = output.at("atom_virial").toTensor().cpu();
    auto atomic_virial = atomic_virial_tensor.accessor<float, 3>();
    for (int ii = 0; ii < ntotal; ii++)
    {
      int i = ilist[ii];
      vatom[i][0] +=  atomic_virial[i][0][0]; // xx
      vatom[i][1] +=  atomic_virial[i][1][1]; // yy 
      vatom[i][2] +=  atomic_virial[i][2][2]; // zz
      vatom[i][3] +=  0.5*(atomic_virial[i][0][1]+atomic_virial[i][1][0]); // xy
      vatom[i][4] +=  0.5*(atomic_virial[i][0][2]+atomic_virial[i][2][0]); // xz
      vatom[i][5] +=  0.5*(atomic_virial[i][1][2]+atomic_virial[i][2][1]); // yz
    }
   }
   
   Will this code work? I haven't tested it yet, but I'd appreciate it if you could tell me what seems to be a problem

Thanks!

**Comment by wcwitt** [comment]
Hi - sorry for the slow reply. How important is this for you? It might make things much slower.

**Reply by pobo95** [reply]
It is my main topic these days, I think parallelism should be a necessity.

**Reply by wcwitt** [reply]
Okay - I will try. Can you say how/why you use the atomic virials (partly for my curiousity)?

**Reply by pobo95** [reply]
To get atomic/local shear and bulk modulus. I'm looking at the correlation between atomic virial and phonon.

**Reply by bernstei** [reply]
BTW, if the atomic virial is the strain derivative of a well-defined atomic energy, then it's also sufficient for calculating the heat flux for Green-Kubo thermal conductivity.

**Reply by pobo95** [reply]
@bernstei Yes you're right! I'm considering thermal conductivity too but what I really want to see is phase transition.

**Reply by nickhine** [reply]
The Green-Kubo thermal conductivity calculation is something we would like to be able to do in our group too, applying it to 2D materials. @pobo95 did you make any progress with this? I saw a PR from you that then got closed - was that code working?

**Reply by pobo95** [reply]
@nickhine 
Yes It works. You can check it in my [repository](https://github.com/pobo95/mace/tree/atom_virial) (atomic virial) and you need this [code](https://github.com/pobo95/pair_mace) to pair with LAMMPS. But it's still in progress, so you can't use it right now, you need more tests. Please feel free to contact me by e-mail.

**Reply by clecust** [reply]
> BTW, if the atomic virial is the strain derivative of a well-defined atomic energy, then it's also sufficient for calculating the heat flux for Green-Kubo thermal conductivity.

Hi, I am very interested to know how to calculate thermal conductivity based on Green-Kubo method without using vflag_atom but vflag_global. Could you please give me an example of the input file for lammps? Thank you very much.
@bernstei

**Reply by bernstei** [reply]
I'm not aware of such a thing being possible. You need vflag_atom, and for the thermal conductivity it also needs to satisfy certain constraints (i.e. that it's the derivative of a well defined local energy)
